File: examples/Clojure/syncpub.clj
(ns syncpub
  (:refer-clojure :exclude [send])
  (:require [zhelpers :as mq]))

;;
;; Synchronized publisher
;;
;; Isaiah Peng <issaria@gmail.com>

(defn -main []
  (let [ctx (mq/context 1)
        publisher (mq/socket ctx mq/pub)
        syncservice (mq/socket ctx mq/rep)
        subscribers-expected 10]
    (mq/bind publisher "tcp://*:5561")
    (mq/bind syncservice "tcp://*:5562")
    ;; Get synchronization from subscribers
    (println "Waiting for subscribers")
    (dotimes [i subscribers-expected]
      ;; - wait for synchronization request
      (mq/recv syncservice)
      ;; - send synchronization reply
      (mq/send syncservice "\u0000"))
    ;; Now broadcast exactly 1M updates followed by END
    (println "Broadicasting messages")
    (dotimes [i 1000000]
      (mq/send publisher "Rhubarb\u0000"))
    (mq/send publisher "END\u0000")
    (.close publisher)
    (.close syncservice)
    (.term ctx)))

File: examples/C++/syncpub.cpp
//
//  Synchronized publisher in C++
//

#include "zhelpers.hpp"

//  We wait for 10 subscribers
#define SUBSCRIBERS_EXPECTED  10

int main () {
	zmq::context_t context(1);

    //  Socket to talk to clients
    zmq::socket_t publisher (context, ZMQ_PUB);

    int sndhwm = 0;
    publisher.setsockopt (ZMQ_SNDHWM, &sndhwm, sizeof (sndhwm));

    publisher.bind("tcp://*:5561");

    //  Socket to receive signals
    zmq::socket_t syncservice (context, ZMQ_REP);
    syncservice.bind("tcp://*:5562");

    //  Get synchronization from subscribers
    int subscribers = 0;
    while (subscribers < SUBSCRIBERS_EXPECTED) {
        
		//  - wait for synchronization request
		s_recv (syncservice);
       
		//  - send synchronization reply
		s_send (syncservice, "");


        subscribers++;
    }
    
    //  Now broadcast exactly 1M updates followed by END
    int update_nbr;
    for (update_nbr = 0; update_nbr < 1000000; update_nbr++) {	
		s_send (publisher, "Rhubarb");
	}
	
    s_send (publisher, "END");

    sleep (1);              //  Give 0MQ time to flush output
    return 0;
}
